### PR TITLE
chore: Move bit-shift overflow checks from Brillig VM into Brillig codegen

### PR DESCRIPTION
**Pull request by guipublic**
# Description

## Problem

Resolves https://github.com/AztecProtocol/aztec-packages/issues/16944

## Summary
Instead of having overflow checks for bit shifts done by the VM, they are now code-gen in the IR.
Brillig bit-shift have now the same semantic as AVM bit-shifts.

## Additional Context
Bit-shift semantic are the same in SSA, ACIR, and Brillig:
Bit shift instructions in SSA include implicit overflow checks (so they are not pure)
They are now expanded into 'pure' bit-shift + explicit overflow check:
- during SSA for ACIR runtime
- during Brillig-gen for unconstrained

We could make the Bit-shift instruction pure in SSA, and avoid the overflow checks both in Brillig and ACIR by adding them directly in Noir:
```
impl Shl for u32 {
  fn shl(self, other) {
   assert(other < 32, "attempt to bit-shift with overflow");
   built_in_shl(self, other);
  }
}
```
The built_in_shl() would layout the pure Shl instruction.

## User Documentation

Check one:
- [ X] No user documentation needed.
- [ ] Changes in _docs/_ included in this PR.
- [ ] **[For Experimental Features]** Changes in _docs/_ to be submitted in a separate PR.

# PR Checklist

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by TomAFrench** [comment]
Can you add some tests in the brillig VM (for evaluate_binary_int_op_shifts) for the various edge-cases? If we're supposed to be zeroing out the result for large shifts then the default rust `>>`\`<<` operators won't behave like that in release mode.

**Comment by guipublic** [comment]
> Can you add some tests in the brillig VM (for evaluate_binary_int_op_shifts) for the various edge-cases? 

There are already 2 tests for the overflow case (that I fixed). I don't see any other meaningful test, so I added some proptest, that will ensure that shifts do not have side effects in the VM.

**Review by TomAFrench — APPROVED** [review]

File: compiler/noirc_evaluator/src/brillig/brillig_gen/brillig_instructions/brillig_binary.rs [hunk 1]
```diff
@@ -113,6 +117,23 @@ impl<Registers: RegisterAllocator> BrilligBlock<'_, Registers> {
...
+        self.brillig_context.binary_instruction(
+            right,
+            *left_bit_size,
+            *right_overflow,
+            BrilligBinaryOp::LessThan,
+        );
+        let msg = "attempt to bit-shift with overflow".to_string();
+        self.brillig_context.codegen_constrain(*right_overflow, Some(msg));
+    }
+
```
Comment (TomAFrench) [inline compiler/noirc_evaluator/src/brillig/brillig_gen/brillig_instructions/brillig_binary.rs:136]: We don't need to pass in `left` here as `right` should have the same bitsize.

Comment (TomAFrench) [reply compiler/noirc_evaluator/src/brillig/brillig_gen/brillig_instructions/brillig_binary.rs:136]: Eh, I guess it doesn't really make much differences at the end of the day and would pick up cases where we've done bad codegen.
